From 63348a12b5b479b51208862dc1623fb56c7da3f2 Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Mon, 5 Oct 2026 09:00:22 +0200
Subject: [PATCH] dnsapi: Use a Unix call to retrieve host information.

wineboot queries the hostname before services are started, so we can't
answer that query in a resolver service. This patch introduces a Unix call
to retrieve the information which will allow us to bypass the service for
hostname queries.

This also avoids delays on startup when the hosts file is very large.
---
 dlls/dnsapi/dnsapi.h    | 10 ++++++
 dlls/dnsapi/libresolv.c | 62 ++++++++++++++++++++++++++++++++++++++
 dlls/dnsapi/main.c      |  4 +--
 dlls/dnsapi/query.c     | 67 +++++++++++++++++++++++++++++++++++++++++
 4 files changed, 141 insertions(+), 2 deletions(-)

diff --git dlls/dnsapi/dnsapi.h dlls/dnsapi/dnsapi.h
index e9c4afa65ab..6eebb1fa343 100644
--- dlls/dnsapi/dnsapi.h
+++ dlls/dnsapi/dnsapi.h
@@ -97,6 +97,7 @@ static inline char *strdup_ua( const char *src )
     return dst;
 }
 
+void init_hostinfo( void );
 void free_host_entries( void );
 void destroy_cache( void );
 BOOL get_cache_data_table( DNS_CACHE_ENTRY ** );
@@ -125,12 +126,21 @@ struct query_params
     DWORD           *len;
 };
 
+struct get_hostinfo_params
+{
+    char        *name;
+    char        *cname;
+    IP4_ADDRESS *ip4;
+    IP6_ADDRESS *ip6;
+};
+
 enum unix_funcs
 {
     unix_get_searchlist,
     unix_get_serverlist,
     unix_set_serverlist,
     unix_query,
+    unix_get_hostinfo,
     unix_funcs_count
 };
 
diff --git dlls/dnsapi/libresolv.c dlls/dnsapi/libresolv.c
index 88706e6f271..dd9b32d00db 100644
--- dlls/dnsapi/libresolv.c
+++ dlls/dnsapi/libresolv.c
@@ -30,6 +30,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/types.h>
+#include <unistd.h>
 
 #ifdef HAVE_NETINET_IN_H
 # include <netinet/in.h>
@@ -324,12 +325,51 @@ static NTSTATUS resolv_query( void *args )
     return ret;
 }
 
+static NTSTATUS get_hostinfo( void *args )
+{
+    struct addrinfo hints = { .ai_flags = AI_CANONNAME }, *result, *next;
+    struct get_hostinfo_params *params = args;
+    struct sockaddr_in *ip4;
+    struct sockaddr_in6 *ip6;
+
+    *params->name  = 0;
+    *params->cname = 0;
+    *params->ip4   = 0;
+    memset( params->ip6, 0, sizeof(*params->ip6) );
+
+    if (gethostname( params->name, DNS_MAX_NAME_LENGTH )) return map_h_errno( h_errno );
+
+    if (!getaddrinfo( params->name, NULL, &hints, &result ))
+    {
+        if (result->ai_canonname) strcpy( params->cname, result->ai_canonname );
+
+        next = result;
+        while (next && next->ai_family != AF_INET) next = next->ai_next;
+        if (next)
+        {
+            ip4 = (struct sockaddr_in *)next->ai_addr;
+            *params->ip4 = ip4->sin_addr.s_addr;
+        }
+        next = result;
+        while (next && next->ai_family != AF_INET6) next = next->ai_next;
+        if (next)
+        {
+            ip6 = (struct sockaddr_in6 *)next->ai_addr;
+            memcpy( params->ip6, ip6->sin6_addr.s6_addr, sizeof(*params->ip6) );
+        }
+        freeaddrinfo( result );
+    }
+
+    return ERROR_SUCCESS;
+}
+
 const unixlib_entry_t __wine_unix_call_funcs[] =
 {
     resolv_get_searchlist,
     resolv_get_serverlist,
     resolv_set_serverlist,
     resolv_query,
+    get_hostinfo,
 };
 
 C_ASSERT( ARRAYSIZE(__wine_unix_call_funcs) == unix_funcs_count );
@@ -397,12 +437,34 @@ static NTSTATUS wow64_resolv_query( void *args )
     return resolv_query( &params );
 }
 
+static NTSTATUS wow64_get_hostinfo( void *args )
+{
+    struct
+    {
+        PTR32 name;
+        PTR32 cname;
+        PTR32 ip4;
+        PTR32 ip6;
+    } const *params32 = args;
+
+    struct get_hostinfo_params params =
+    {
+        ULongToPtr(params32->name),
+        ULongToPtr(params32->cname),
+        ULongToPtr(params32->ip4),
+        ULongToPtr(params32->ip6)
+    };
+
+    return get_hostinfo( &params );
+}
+
 const unixlib_entry_t __wine_unix_call_wow64_funcs[] =
 {
     wow64_resolv_get_searchlist,
     wow64_resolv_get_serverlist,
     resolv_set_serverlist,
     wow64_resolv_query,
+    wow64_get_hostinfo,
 };
 
 C_ASSERT( ARRAYSIZE(__wine_unix_call_wow64_funcs) == unix_funcs_count );
diff --git dlls/dnsapi/main.c dlls/dnsapi/main.c
index d11ed320c48..e513e023d82 100644
--- dlls/dnsapi/main.c
+++ dlls/dnsapi/main.c
@@ -40,8 +40,8 @@ BOOL WINAPI DllMain( HINSTANCE hinst, DWORD reason, void *reserved )
     {
     case DLL_PROCESS_ATTACH:
         DisableThreadLibraryCalls( hinst );
-        if (__wine_init_unix_call())
-            ERR( "No libresolv support, expect problems\n" );
+        if (!__wine_init_unix_call()) init_hostinfo();
+        else ERR( "No libresolv support, expect problems\n" );
         break;
     case DLL_PROCESS_DETACH:
         if (reserved) break;
diff --git dlls/dnsapi/query.c dlls/dnsapi/query.c
index b8f5a9d096d..01b9df885ce 100644
--- dlls/dnsapi/query.c
+++ dlls/dnsapi/query.c
@@ -797,6 +797,68 @@ static DNS_STATUS do_query_dns( const char *name, WORD type, DWORD options, void
     return ret;
 }
 
+static struct
+{
+    char name[DNS_MAX_NAME_LENGTH];
+    char cname[DNS_MAX_NAME_LENGTH];
+    IP4_ADDRESS ip4;
+    IP6_ADDRESS ip6;
+} hostinfo;
+
+void init_hostinfo( void )
+{
+    struct get_hostinfo_params params;
+
+    params.name  = hostinfo.name;
+    params.cname = hostinfo.cname;
+    params.ip4   = &hostinfo.ip4;
+    params.ip6   = &hostinfo.ip6;
+    RESOLV_CALL( get_hostinfo, &params );
+}
+
+static DNS_STATUS do_query_hostname( const char *name, WORD type, DNS_RECORDA **result )
+{
+    static const IP6_ADDRESS ip6_zero;
+    DNS_RRSET rrset;
+    DNS_RECORDA *rec;
+
+    DNS_RRSET_INIT( rrset );
+    switch (type)
+    {
+    case DNS_TYPE_A:
+        if (!hostinfo.ip4) return DNS_ERROR_RCODE_NAME_ERROR;
+        if (*hostinfo.cname)
+        {
+            if ((rec = create_cname_record( hostinfo.name, hostinfo.cname ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+            if ((rec = create_a_record( hostinfo.cname, &hostinfo.ip4 ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+        }
+        else if ((rec = create_a_record( hostinfo.name, &hostinfo.ip4 ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+        break;
+
+    case DNS_TYPE_AAAA:
+        if (!memcmp(&hostinfo.ip6, &ip6_zero, sizeof(ip6_zero))) return DNS_ERROR_RCODE_NAME_ERROR;
+        if (*hostinfo.cname)
+        {
+            if ((rec = create_cname_record( hostinfo.name, hostinfo.cname ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+            if ((rec = create_aaaa_record( hostinfo.cname, &hostinfo.ip6 ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+        }
+        else if ((rec = create_aaaa_record( hostinfo.name, &hostinfo.ip6 ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+        break;
+
+    case DNS_TYPE_CNAME:
+        if (!*hostinfo.cname) return DNS_ERROR_RCODE_NAME_ERROR;
+        if ((rec = create_cname_record( hostinfo.name, hostinfo.cname ))) DNS_RRSET_ADD( rrset, (DNS_RECORD *)rec );
+        break;
+
+    default:
+        return DNS_ERROR_RCODE_NOT_IMPLEMENTED;
+    }
+    DNS_RRSET_TERMINATE( rrset );
+
+    *result = (DNS_RECORDA *)rrset.pFirstRR;
+    return ERROR_SUCCESS;
+}
+
 /******************************************************************************
  * DnsQuery_UTF8              [DNSAPI.@]
  *
@@ -851,6 +913,11 @@ DNS_STATUS WINAPI DnsQuery_UTF8( const char *name, WORD type, DWORD options, voi
 
     if ((ret = DnsValidateName_UTF8( name, DnsNameDomain )) && ret != DNS_ERROR_NON_RFC_NAME) return ret;
 
+    if ((type == DNS_TYPE_A || type == DNS_TYPE_AAAA || type == DNS_TYPE_CNAME) && !strcasecmp( name, hostinfo.name ))
+    {
+        return do_query_hostname( name, type, result );
+    }
+
     if ((type == DNS_TYPE_A || type == DNS_TYPE_AAAA || type == DNS_TYPE_CNAME) && !(options & DNS_QUERY_NO_HOSTS_FILE))
     {
         if (!(options & DNS_QUERY_BYPASS_CACHE) && !find_cache_result( SERVICE_HOSTS, name, type, result ))
-- 
GitLab

